dali/operators/generic/roi_random_crop.cc

```diff
@@ -0,0 +1,226 @@
+    DALI_ENFORCE((in_shape_arg_.IsDefined() + (ws.NumInput() == 1)) == 1,
+      "``in_shape`` argument is incompatible with providing an input.");
+    if (in_shape_arg_.IsDefined()) {
+      in_shape_.resize(nsamples, ndim);
+      in_shape_arg_.Acquire(spec_, ws, nsamples, sh);
+      for (int s = 0; s < nsamples; s++) {
+        auto sample_sh = in_shape_.tensor_shape_span(s);
+        for (int d = 0; d < ndim; d++) {
+          sample_sh[d] = in_shape_arg_[s].data[d];
+        }
+      }
+    } else {
+      auto &in = ws.template InputRef<CPUBackend>(0);
+      in_shape_ = in.shape();
+    }
+
+    for (int s = 0; s < nsamples; s++) {
+      auto sample_sh = in_shape_.tensor_shape_span(s);
+      for (int d = 0; d < ndim; d++) {
+        DALI_ENFORCE(sample_sh[d] >= 0,
+                     make_string("Input shape can't be negative. Got ",
+                                 sample_sh[d], " for d=", d));
+        DALI_ENFORCE(crop_shape_[s].data[d] >= 0,
+                     make_string("Crop shape can't be negative. Got ", crop_shape_[s].data[d],
+                                 " for d=", d));
+        DALI_ENFORCE(sample_sh[d] >= crop_shape_[s].data[d],
+                     make_string("Cropping shape can't be bigger than the input shape. "
+                                 "Got: crop_shape[", crop_shape_[s].data[d], "] and sample_shape[",
+                                 sample_sh[d], "] for d=", d));
+      }
+      if (roi_shape_.IsDefined()) {
+        for (int d = 0; d < ndim; d++) {
+          auto roi_end = roi_start_[s].data[d] + roi_shape_[s].data[d];
+          DALI_ENFORCE(roi_start_[s].data[d] >= 0 && sample_sh[d] >= roi_end,
+                       make_string("ROI can't be out of bounds. Got roi_start[",
+                                   roi_start_[s].data[d], "], roi_end[", roi_end,
+                                   "], sample_shape[", sample_sh[d], "], for d=", d));
+        }
+      } else {
+        for (int d = 0; d < ndim; d++) {
+          DALI_ENFORCE(roi_start_[s].data[d] >= 0 && sample_sh[d] >= roi_end_[s].data[d],
+                       make_string("ROI can't be out of bounds. Got roi_start[",
+                                   roi_start_[s].data[d], "], roi_end[", roi_end_[s].data[d],
+                                   "], sample_shape[", sample_sh[d], "], for d=", d));
+        }
+      }
+    }
+  }
+
+  output_desc.resize(1);
+  output_desc[0].shape = uniform_list_shape(nsamples, sh);
+  output_desc[0].type = TypeTable::GetTypeInfo(DALI_INT64);
+  return true;
+}
+
+void ROIRandomCropCPU::RunImpl(workspace_t<CPUBackend> &ws) {
+  auto &out_crop_start = ws.template OutputRef<CPUBackend>(0);
+  auto crop_start = view<int64_t, 1>(out_crop_start);
+
+  int nsamples = crop_start.shape.size();
+  int ndim = crop_start[0].shape[0];
+
+  auto& thread_pool = ws.GetThreadPool();
+
+  for (int sample_idx = 0; sample_idx < nsamples; sample_idx++) {
+    thread_pool.AddWork(
```

Comment:
I wonder if asynchronous execution is justified here - after all this operator performs `ndim` loops per sample - not much compared to the overhead of scheduling work to thread pool.

Reply:
I can get rid of it